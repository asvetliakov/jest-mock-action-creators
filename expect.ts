import { ActionCreatorInvokation } from "./";
import diff = require("jest-diff");

declare global {
    namespace jest {
        interface Matchers<R> {
            toBeCalledWithActionCreator(actionCreatorName: string | Function, ...args: any[]): void;
        }
    }
}

expect.extend({
    toBeCalledWithActionCreator(received: jest.Mock<any>, actionCreatorName: string | Function, ...args: any[]): { message(): string, pass: boolean } {
        if (!received || !received.mock) {
            throw new Error(
                this.utils.matcherHint("[.not].toBeCalledWithActionCreator", "dispatch", "") +
                "\n\n" +
                `${(this.utils.RECEIVED_COLOR as any)("dispatch")} value must be a mock function or spy.\n` +
                this.utils.printWithType("Received", received, val => (this.utils.RECEIVED_COLOR as any)(this.utils.stringify(val)))
            );
        }

        if (typeof actionCreatorName !== "string" && typeof actionCreatorName !== "function") {
            throw new Error(
                this.utils.matcherHint("[.not].toBeCalledWithActionCreator", "dispatch", "actionCreator") +
                "\n\n" +
                `${(this.utils.EXPECTED_COLOR as any)("actionCreator")} must be a string or function.\n` +
                "Received: " + this.utils.printReceived(typeof actionCreatorName)
            );
        }

        const creatorName = typeof actionCreatorName === "function" ? actionCreatorName.name : actionCreatorName;

        const call = received.mock.calls.find((val: any[]) => {
            return val && val[0] && val[0].actionCreator === creatorName
        });
        const invokation: ActionCreatorInvokation | undefined = call ? call[0] : undefined;

        let pass = true;
        if (!invokation) {
            pass = false;
        } else if (args.length > 0) {
            try {
                expect(args).toEqual(invokation.args);
            } catch {
                pass = false;
            }
        }

        const formatExpected = (): string => {
            let msg = `Expected dispatch() ${this.utils.printExpected(pass ? "to not call" : "to call")} action creator ${this.utils.printExpected(creatorName)}\n`;
            if (args.length > 0) {
                msg += `  With arguments: ${(this.utils.EXPECTED_COLOR as any)(this.utils.stringify(args))}\n`;
            }
            return msg;
        }

        const formatReceived = (): string => {
            if (!pass && !invokation) {
                return `${(this.utils.RECEIVED_COLOR as any)("But it never called it.")}`;
            }
            if (!pass && invokation) {
                if (invokation.args && invokation.args.length > 0) {
                    let msg =  `${(this.utils.RECEIVED_COLOR as any)("But it was called it with arguments: " + this.utils.stringify(invokation.args))}`;
                    if (args.length > 0) {
                        const diffStr = diff(args, invokation.args, { expand: (this as any).expand });
                        if (diffStr) {
                            msg += `\n\nDifference\n\n${diffStr}`;
                        }
                    }
                    return msg;
                } else {
                    return `${(this.utils.RECEIVED_COLOR as any)("But it was called it")}`;
                }
            }
            return `${(this.utils.RECEIVED_COLOR as any)("But it was called it")}`;
        }

        const message = pass
            ? () => this.utils.matcherHint(".not.toBeCalledWithActionCreator", "dispatch", creatorName) + "\n\n" +
                `${formatExpected()}\n` +
                `  ${formatReceived()}`
            : () => this.utils.matcherHint(".toBeCalledWithActionCreator", "dispatch", creatorName) + "\n\n" +
                `${formatExpected()}\n` +
                `  ${formatReceived()}`;
        return {
            message,
            pass
        };
    }
});