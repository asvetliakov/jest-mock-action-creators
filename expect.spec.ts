import { mockActionCreators } from "./";
import "./expect";

it("Throws if calling with non mocked dispatch", () => {
    const f = () => { };

    expect(() => expect(f).toBeCalledWithActionCreator("abc")).toThrowErrorMatchingSnapshot();
});

it("Throws if action creator name is not a string or function", () => {
    const dis = jest.fn();
    expect(() => expect(dis).toBeCalledWithActionCreator(true as any)).toThrowErrorMatchingSnapshot();
    expect(() => expect(dis).toBeCalledWithActionCreator({} as any)).toThrowErrorMatchingSnapshot();
});

describe("Checks invokations", () => {
    let dis: jest.Mock<any>;
    let ac1: jest.Mock<any>;
    let ac2: jest.Mock<any>;
    beforeEach(() => {
        ac1 = jest.fn();
        Object.defineProperty(ac1, "name", { value: "ac1" });
        ac2 = jest.fn();
        Object.defineProperty(ac2, "name", { value: "ac2" });

        mockActionCreators(ac1, ac2);

        dis = jest.fn();

        dis(ac1("a", "b", true));
    });

    it("Checks for just name without arguments", () => {
        expect(dis).toBeCalledWithActionCreator(ac1);
        expect(dis).not.toBeCalledWithActionCreator(ac2);
        expect(() => expect(dis).not.toBeCalledWithActionCreator(ac1)).toThrowErrorMatchingSnapshot();
        expect(() => expect(dis).toBeCalledWithActionCreator(ac2)).toThrowErrorMatchingSnapshot();
    });

    it("Checks for name with arguments", () => {
        expect(dis).toBeCalledWithActionCreator(ac1, "a", "b", true);
        expect(() => expect(dis).toBeCalledWithActionCreator(ac1, "e", "k", false)).toThrowErrorMatchingSnapshot();
        expect(dis).not.toBeCalledWithActionCreator(ac1, "e", "k");

        expect(dis).not.toBeCalledWithActionCreator(ac2, "a", "b", true);
        expect(() => expect(dis).toBeCalledWithActionCreator(ac2, "a", "b", true)).toThrowErrorMatchingSnapshot();
    });
})
