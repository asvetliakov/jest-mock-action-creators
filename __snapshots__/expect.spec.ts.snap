// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checks invokations Checks for just name without arguments 1`] = `
"[2mexpect([22m[31mdispatch[39m[2m).not.toBeCalledWithActionCreator([22m[32mac1[39m[2m)[22m

Expected dispatch() [32m\\"to not call\\"[39m action creator [32m\\"ac1\\"[39m

  [31mBut it was called it[39m"
`;

exports[`Checks invokations Checks for just name without arguments 2`] = `
"[2mexpect([22m[31mdispatch[39m[2m).toBeCalledWithActionCreator([22m[32mac2[39m[2m)[22m

Expected dispatch() [32m\\"to call\\"[39m action creator [32m\\"ac2\\"[39m

  [31mBut it never called it.[39m"
`;

exports[`Checks invokations Checks for name with arguments 1`] = `
"[2mexpect([22m[31mdispatch[39m[2m).toBeCalledWithActionCreator([22m[32mac1[39m[2m)[22m

Expected dispatch() [32m\\"to call\\"[39m action creator [32m\\"ac1\\"[39m
  With arguments: [32m[\\"e\\", \\"k\\", false][39m

  [31mBut it was called it with arguments: [\\"a\\", \\"b\\", true][39m

Difference

[32m- Expected[39m
[31m+ Received[39m

[2m  Array [[22m
[32m-   \\"e\\",[39m
[32m-   \\"k\\",[39m
[32m-   false,[39m
[31m+   \\"a\\",[39m
[31m+   \\"b\\",[39m
[31m+   true,[39m
[2m  ][22m"
`;

exports[`Checks invokations Checks for name with arguments 2`] = `
"[2mexpect([22m[31mdispatch[39m[2m).toBeCalledWithActionCreator([22m[32mac2[39m[2m)[22m

Expected dispatch() [32m\\"to call\\"[39m action creator [32m\\"ac2\\"[39m
  With arguments: [32m[\\"a\\", \\"b\\", true][39m

  [31mBut it never called it.[39m"
`;

exports[`Throws if action creator name is not a string or function 1`] = `
"[2mexpect([22m[31mdispatch[39m[2m)[.not].toBeCalledWithActionCreator([22m[32mactionCreator[39m[2m)[22m

[32mactionCreator[39m must be a string or function.
Received: [31m\\"boolean\\"[39m"
`;

exports[`Throws if action creator name is not a string or function 2`] = `
"[2mexpect([22m[31mdispatch[39m[2m)[.not].toBeCalledWithActionCreator([22m[32mactionCreator[39m[2m)[22m

[32mactionCreator[39m must be a string or function.
Received: [31m\\"object\\"[39m"
`;

exports[`Throws if calling with non mocked dispatch 1`] = `
"[2mexpect([22m[31mdispatch[39m[2m)[.not].toBeCalledWithActionCreator([22m[2m)[22m

[31mdispatch[39m value must be a mock function or spy.
Received:
  function: [31m[Function f][39m"
`;
